plugin/forward/forward.go

```diff
@@ -68,6 +72,14 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 		return plugin.NextOrFailure(f.Name(), f.Next, ctx, w, r)
 	}
 
+	if f.maxQueryCount > 0 {
+		count := atomic.AddInt64(&(f.queryCount), 1)
+		defer atomic.AddInt64(&(f.queryCount), -1)
+		if count > f.maxQueryCount {
```

Comment:
In theory there is a race condition where we may drop some extra queries when we're hovering near the edge. I don't think we care though, not enough to mutex this whole thing.

Reply:
Yes, I lingered on this issue for a bit after seeing it happen during testing, but decided to stick with simple/performant atomic despite this.  The count can go over the limit if overwhelmed, since we don't check for the limit in the same atomic action as the increment.  Would be nice if there was an atomic action "increment-if-less-than", but I don't think this is a big deal, since these increments above the limit are immediately decremented.

Adding an atomic check of the value before the increment lessens, but does not completely eliminate the issue.  But I don't think it's a problem enough to warrant the extra check.